File: scripts/fill_competitor_codes.py
"""
fill_competitor_codes.py — Poblar competitor_codes desde brand_crossrefs de los JSON Donaldson.

Conecta DIRECTAMENTE a PostgreSQL (sin pasar por el API server de Render)
para actualizar solo los productos con competitor_codes vacío.

Uso:
    python3 fill_competitor_codes.py [--dry-run]

Requiere: pip install psycopg2-binary
"""

import argparse
import glob
import json
import logging
import os

import psycopg2
import psycopg2.extras

logging.basicConfig(level=logging.INFO, format="%(levelname)s %(message)s")

DB = {
    "host":     "ballast.proxy.rlwy.net",
    "port":     18263,
    "dbname":   "railway",
    "user":     "postgres",
    "password": "qUiKsOlOyDSyHZogyqhhxTTPlAuuLEkm",
    "sslmode":  "require",
}
BATCH   = 200
SCRIPTS = os.path.dirname(os.path.abspath(__file__))


def load_all_json():
    """Carga todos los donaldson_*_results.json y extrae (sku, competitor_codes)."""
    rows = {}  # sku → competitor_codes list
    files = sorted(glob.glob(os.path.join(SCRIPTS, "donaldson_*_results.json")))
    for fpath in files:
        cat = os.path.basename(fpath).replace("donaldson_", "").replace("_results.json", "")
        try:
            data = json.load(open(fpath, encoding="utf-8"))
        except Exception as e:
            logging.warning(f"[{cat}] No se pudo leer {fpath}: {e}")
            continue

        total = len(data)
        with_br = 0
        for p in data:
            br = p.get("brand_crossrefs") or {}
            if not br:
                continue
            sku = p.get("sku_elimfilters")
            if not sku:
                continue
            codes = []
            for brand, parts in br.items():
                for pn in (parts or []):
                    if brand and pn:
                        codes.append({"manufacturer": brand.upper(), "code": str(pn).upper()})
            if codes:
                rows[sku] = codes
                with_br += 1

        logging.info(f"[{cat:20}] {total:5} productos | {with_br:5} con brand_crossrefs")

    return rows


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true", help="Solo muestra stats, sin escribir al DB")
    args = parser.parse_args()

    rows_map = load_all_json()
    rows = [(sku, json.dumps(codes)) for sku, codes in rows_map.items()]
    logging.info(f"\nTotal productos con competitor_codes para actualizar: {len(rows)}")

    if args.dry_run:
        logging.info("DRY RUN — sin escritura al DB")
        if rows:
            logging.info(f"Ejemplo: {rows[0][0]} → {json.loads(rows[0][1])[:3]}...")
        return

    conn = psycopg2.connect(**DB)
    conn.autocommit = False
    cur  = conn.cursor()

    total_updated = 0
    for i in range(0, len(rows), BATCH):
        batch = rows[i:i + BATCH]
        psycopg2.extras.execute_batch(cur, """
            UPDATE elimfilters_catalog
            SET    competitor_codes = %s::jsonb
            WHERE  sku = %s
              AND  (competitor_codes IS NULL
                    OR jsonb_array_length(COALESCE(competitor_codes,'[]'::jsonb)) = 0)
        """, [(codes, sku) for sku, codes in batch])
        total_updated += cur.rowcount
        conn.commit()
        pct = (i + len(batch)) / len(rows) * 100
        logging.info(f"  Lote {i//BATCH+1}: {cur.rowcount} actualizados | {pct:.0f}%")

    cur.close()
    conn.close()
    logging.info(f"\nDONE: {total_updated} productos actualizados")


if __name__ == "__main__":
    main()

File: scripts/fill_competitor_codes.js
/**
 * fill_competitor_codes.js
 * Popula competitor_codes desde brand_crossrefs de los JSON Donaldson.
 * Usa pg (ya instalado en Render) y DATABASE_URL env var.
 *
 * Uso en Render shell:
 *   node scripts/fill_competitor_codes.js [--dry-run]
 */

const { Pool } = require('pg');
const fs   = require('fs');
const path = require('path');

const DRY_RUN = process.argv.includes('--dry-run');
const BATCH   = 100;
// JSON files live in the same directory as this script
const DIR     = process.cwd();

const DB_URL = process.env.DATABASE_URL;

const pool = DRY_RUN ? null : (DB_URL
  ? new Pool({ connectionString: DB_URL, ssl: { rejectUnauthorized: false } })
  : new Pool({
      connectionString: process.env.LEGACY_DB_URL,
      ssl:      { rejectUnauthorized: false },
    })
);

function loadAllJson() {
  const rows = {};
  const files = fs.readdirSync(DIR)
    .filter(f => /^donaldson_.*_results\.json$/.test(f))
    .map(f => path.join(DIR, f))
    .sort();

  for (const fpath of files) {
    const cat = path.basename(fpath).replace('donaldson_', '').replace('_results.json', '');
    let data;
    try { data = JSON.parse(fs.readFileSync(fpath, 'utf8')); }
    catch (e) { console.warn(`[${cat}] No se pudo leer: ${e.message}`); continue; }

    let withBr = 0;
    for (const p of data) {
      const br  = p.brand_crossrefs || {};
      const sku = p.sku_elimfilters;
      if (!sku || !Object.keys(br).length) continue;

      const codes = [];
      for (const [brand, parts] of Object.entries(br)) {
        for (const pn of (parts || [])) {
          if (brand && pn) codes.push({ manufacturer: brand.toUpperCase(), code: String(pn).toUpperCase() });
        }
      }
      if (codes.length) { rows[sku] = codes; withBr++; }
    }
    console.log(`[${cat.padEnd(22)}] ${data.length} productos | ${withBr} con brand_crossrefs`);
  }
  return rows;
}

async function main() {
  const rowsMap = loadAllJson();
  const rows    = Object.entries(rowsMap); // [[sku, codes], ...]
  console.log(`\nTotal a actualizar: ${rows.length}`);

  if (DRY_RUN) {
    console.log('DRY RUN — sin escritura');
    if (rows.length) console.log('Ejemplo:', rows[0][0], '→', rows[0][1].slice(0, 3));
    return;
  }

  const client = await pool.connect();
  let totalUpdated = 0;

  try {
    for (let i = 0; i < rows.length; i += BATCH) {
      const batch = rows.slice(i, i + BATCH);
      let batchUpdated = 0;

      await client.query('BEGIN');
      for (const [sku, codes] of batch) {
        const res = await client.query(`
          UPDATE elimfilters_catalog
          SET    competitor_codes = $1::jsonb
          WHERE  sku = $2
            AND  (competitor_codes IS NULL
                  OR jsonb_array_length(COALESCE(competitor_codes,'[]'::jsonb)) = 0)
        `, [JSON.stringify(codes), sku]);
        batchUpdated += res.rowCount;
      }
      await client.query('COMMIT');

      totalUpdated += batchUpdated;
      const pct = Math.round((i + batch.length) / rows.length * 100);
      console.log(`  Lote ${Math.floor(i/BATCH)+1}: ${batchUpdated} actualizados | ${pct}%`);
    }
  } catch (err) {
    await client.query('ROLLBACK');
    console.error('ERROR — ROLLBACK:', err.message);
    process.exit(1);
  } finally {
    client.release();
    await pool.end();
  }

  console.log(`\nDONE: ${totalUpdated} productos actualizados`);
}

main().catch(err => { console.error(err); process.exit(1); });
